paddle/fluid/operators/run_program_op.h

```diff
@@ -196,10 +202,45 @@ static void AppendSafeEagerDeletionSkipVars(
   }
 }
 
+static ExecutorInfo GetExecutorInfoFromCache(
+    std::unordered_map<framework::ProgramDesc *, ExecutorInfo> *cached_exe_info,
+    const framework::ExecutionContext &ctx,
+    const std::vector<std::vector<std::string>> &ctx_output_names,
+    bool is_grad = false) {
+  auto *program = ctx.Attr<BlockDesc *>("global_block")->Program();
+
+  auto it = cached_exe_info->find(program);
+  if (it == cached_exe_info->end()) {
+    VLOG(1) << "create exe_info for program: " << program;
+    auto exe = std::make_shared<framework::Executor>(ctx.GetPlace());
+    // skip delete vars
+    std::vector<std::string> skip_vars;
+    for (auto &output_names : ctx_output_names) {
+      details::AppendSkipDeletionVars(output_names, &skip_vars);
+    }
+
+    if (is_grad) {
+      details::AppendSafeEagerDeletionSkipVars(*program, &skip_vars);
+    }
+
+    VLOG(2) << "Prepare to skip " << skip_vars.size()
+            << " var(s): " << string::join_strings(skip_vars, ' ');
+    std::shared_ptr<framework::ExecutorPrepareContext> exe_ctx =
+        std::move(exe->Prepare(*program, 0, skip_vars));
+    cached_exe_info->emplace(program, std::make_tuple(exe, exe_ctx));
+  }
+
+  return cached_exe_info->at(program);
+}
+
 }  // namespace details
 
 template <typename DeviceContext, typename T>
 class RunProgramOpKernel : public framework::OpKernel<T> {
+ private:
+  static std::unordered_map<framework::ProgramDesc *, ExecutorInfo>
```

Comment:
`static` may introduce problems,maybe singleton is better
![image](https://user-images.githubusercontent.com/22561442/100190679-c29a7880-2f29-11eb-9930-eb77f04d8b18.png)

Reply:
Thanks advice~~, modified into singleton